File: app/src/main/java/wangdaye/com/geometricweather/theme/weatherView/materialWeatherView/implementor/SnowImplementor.java
package wangdaye.com.geometricweather.theme.weatherView.materialWeatherView.implementor;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;

import androidx.annotation.ColorInt;
import androidx.annotation.IntDef;
import androidx.annotation.Size;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.ColorUtils;

import java.util.Random;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.theme.weatherView.materialWeatherView.MaterialWeatherView;

/**
 * Snow implementor.
 * */

public class SnowImplementor extends MaterialWeatherView.WeatherAnimationImplementor {

    private final Paint mPaint;
    private final Snow[] mSnows;

    private float mLastDisplayRate;

    private float mLastRotation3D;
    private static final float INITIAL_ROTATION_3D = 1000;

    private static final int SNOW_COUNT = 90;

    @ColorInt
    private int mBackgroundColor;

    public static final int TYPE_SNOW_DAY = 1;
    public static final int TYPE_SNOW_NIGHT = 2;

    @IntDef({TYPE_SNOW_DAY, TYPE_SNOW_NIGHT})
    @interface TypeRule {}

    private static class Snow {

        private float mCX;
        private float mCY;

        float centerX;
        float centerY;
        float radius;

        @ColorInt
        int color;
        float scale;

        float speedX;
        float speedY;

        private final int mViewWidth;
        private final int mViewHeight;

        private final int mCanvasSize;

        private Snow(int viewWidth, int viewHeight, @ColorInt int color, float scale) {
            mViewWidth = viewWidth;
            mViewHeight = viewHeight;

            mCanvasSize = (int) Math.pow(
                    viewWidth * viewWidth + viewHeight * viewHeight,
                    0.5
            );

            this.radius = (float) (
                    mCanvasSize * (
                            0.005 + new Random().nextDouble() * 0.007
                    ) * scale
            );

            this.color = color;
            this.scale = scale;

            this.speedY = (float) (
                    mCanvasSize / (
                            1000.0 * (2.5 + new Random().nextDouble())
                    ) * 3.0
            );

            init(true);
        }

        private void init(boolean firstTime) {
            Random r = new Random();
            mCX = r.nextInt(mCanvasSize);
            if (firstTime) {
                mCY = r.nextInt((int) (mCanvasSize - radius)) - mCanvasSize;
            } else {
                mCY = -radius;
            }
            speedX = r.nextInt((int) (2 * speedY)) - speedY;

            computeCenterPosition();
        }

        private void computeCenterPosition() {
            centerX = (int) (mCX - (mCanvasSize - mViewWidth) * 0.5);
            centerY = (int) (mCY - (mCanvasSize - mViewHeight) * 0.5);
        }

        void move(long interval, float deltaRotation3D) {
            mCX += speedX * interval * Math.pow(scale, 1.5);
            mCY += speedY * interval * (Math.pow(scale, 1.5) - 5 * Math.sin(deltaRotation3D * Math.PI / 180.0));

            if (centerY >= mCanvasSize) {
                init(false);
            } else {
                computeCenterPosition();
            }
        }
    }

    public SnowImplementor(@Size(2) int[] canvasSizes, @TypeRule int type) {
        mPaint = new Paint();
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setAntiAlias(true);

        int[] colors = new int[3];
        switch (type) {
            case TYPE_SNOW_DAY:
                mBackgroundColor = Color.rgb(104, 186, 255);
                colors = new int[] {
                        Color.rgb(128, 197, 255),
                        Color.rgb(185, 222, 255),
                        Color.rgb(255, 255, 255)};
                break;

            case TYPE_SNOW_NIGHT:
                mBackgroundColor = Color.rgb(26, 91, 146);
                colors = new int[] {
                        Color.rgb(40, 102, 155),
                        Color.rgb(99, 144, 182),
                        Color.rgb(255, 255, 255)};
                break;
        }
        float[] scales = new float[] {0.6F, 0.8F, 1};

        mSnows = new Snow[SNOW_COUNT];
        for (int i = 0; i < mSnows.length; i ++) {
            mSnows[i] = new Snow(
                    canvasSizes[0],
                    canvasSizes[1],
                    colors[i * 3 / mSnows.length],
                    scales[i * 3 / mSnows.length]
            );
        }

        mLastDisplayRate = 0;
        mLastRotation3D = INITIAL_ROTATION_3D;
    }

    @Override
    public void updateData(@Size(2) int[] canvasSizes, long interval,
                           float rotation2D, float rotation3D) {
        for (Snow s : mSnows) {
            s.move(interval, mLastRotation3D == INITIAL_ROTATION_3D ? 0 : rotation3D - mLastRotation3D);
        }
        mLastRotation3D = rotation3D;
    }

    @Override
    public void draw(@Size(2) int[] canvasSizes, Canvas canvas,
                     float displayRate, float scrollRate, float rotation2D, float rotation3D) {

        if (displayRate >= 1) {
            canvas.drawColor(mBackgroundColor);
        } else {
            canvas.drawColor(
                    ColorUtils.setAlphaComponent(
                            mBackgroundColor,
                            (int) (displayRate * 255)));
        }

        if (scrollRate < 1) {
            canvas.rotate(
                    rotation2D,
                    canvasSizes[0] * 0.5F,
                    canvasSizes[1] * 0.5F);
            for (Snow s : mSnows) {
                mPaint.setColor(s.color);
                if (displayRate < mLastDisplayRate) {
                    mPaint.setAlpha((int) (displayRate * (1 - scrollRate) * 255));
                } else {
                    mPaint.setAlpha((int) ((1 - scrollRate) * 255));
                }
                canvas.drawCircle(s.centerX, s.centerY, s.radius, mPaint);
            }
        }

        mLastDisplayRate = displayRate;
    }

    @ColorInt
    public static int getThemeColor(Context context, @TypeRule int type) {
        switch (type) {
            case TYPE_SNOW_DAY:
                return Color.rgb(104, 186, 255);

            case TYPE_SNOW_NIGHT:
                return Color.rgb(26, 91, 146);
        }
        return ContextCompat.getColor(context, R.color.colorPrimary);
    }
}

File: ui-weather-view/src/main/java/org/breezyweather/ui/theme/weatherView/materialWeatherView/implementor/SnowImplementor.kt
/*
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.theme.weatherView.materialWeatherView.implementor

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import androidx.annotation.ColorInt
import androidx.annotation.Size
import org.breezyweather.ui.theme.weatherView.materialWeatherView.MaterialWeatherView.WeatherAnimationImplementor
import java.util.Random
import kotlin.math.pow
import kotlin.math.sin

/**
 * Snow implementor.
 */
class SnowImplementor(
    @Size(2) canvasSizes: IntArray,
    animate: Boolean,
    daylight: Boolean,
) : WeatherAnimationImplementor() {
    private val mAnimate = animate
    private val mPaint = Paint().apply {
        style = Paint.Style.FILL
        isAntiAlias = true
    }
    private val mSnows: Array<Snow>
    private var mLastRotation3D: Float

    private class Snow(
        private val mViewWidth: Int,
        private val mViewHeight: Int,
        @field:ColorInt @param:ColorInt val color: Int,
        val scale: Float,
    ) {
        private var mCX = 0f
        private var mCY = 0f
        var centerX = 0f
        var centerY = 0f
        var radius: Float
        var speedX = 0f
        var speedY: Float
        private val mCanvasSize: Int

        init {
            mCanvasSize = (mViewWidth * mViewWidth + mViewHeight * mViewHeight).toDouble().pow(0.5).toInt()
            radius = (mCanvasSize * (0.005 + Random().nextDouble() * 0.007) * scale).toFloat()
            speedY = (mCanvasSize / (1000.0 * (2.5 + Random().nextDouble())) * SNOW_SPEED).toFloat()
            init(true)
        }

        private fun init(firstTime: Boolean) {
            val r = Random()
            mCX = r.nextInt(mCanvasSize).toFloat()
            mCY = if (firstTime) {
                (r.nextInt((mCanvasSize - radius).toInt()) - mCanvasSize).toFloat()
            } else {
                -radius
            }

            val bound = (2 * speedY).toInt()
            // Fix bound to at least two to prevent crashes on small screens (where bound would be 0) and provide
            // some variation on medium screens (where bound would always be 1 and the RNG would only give 0)
            speedX = r.nextInt(if (bound >= 2) bound else 2) - speedY
            computeCenterPosition()
        }

        private fun computeCenterPosition() {
            centerX = (mCX - (mCanvasSize - mViewWidth) * 0.5).toInt().toFloat()
            centerY = (mCY - (mCanvasSize - mViewHeight) * 0.5).toInt().toFloat()
        }

        fun move(interval: Long, deltaRotation3D: Float) {
            mCX += (speedX * interval * scale.toDouble().pow(1.5)).toFloat()
            mCY +=
                (speedY * interval * (scale.toDouble().pow(1.5) - 5 * sin(deltaRotation3D * Math.PI / 180.0))).toFloat()
            if (centerY >= mCanvasSize) {
                init(false)
            } else {
                computeCenterPosition()
            }
        }
    }

    init {
        val colors = if (daylight) {
            intArrayOf(
                Color.rgb(190, 225, 255),
                Color.rgb(211, 233, 255),
                Color.rgb(255, 255, 255)
            )
        } else {
            intArrayOf(
                Color.rgb(111, 133, 155),
                Color.rgb(140, 161, 182),
                Color.rgb(255, 255, 255)
            )
        }
        val scales = floatArrayOf(0.6f, 0.8f, 1f)
        mSnows = Array(SNOW_COUNT) { i ->
            Snow(
                canvasSizes[0],
                canvasSizes[1],
                colors[i * 3 / SNOW_COUNT],
                scales[i * 3 / SNOW_COUNT]
            )
        }
        mLastRotation3D = INITIAL_ROTATION_3D
    }

    override fun updateData(
        @Size(2) canvasSizes: IntArray,
        interval: Long,
        rotation2D: Float,
        rotation3D: Float,
    ) {
        for (s in mSnows) {
            s.move(interval, if (mLastRotation3D == INITIAL_ROTATION_3D) 0f else rotation3D - mLastRotation3D)
        }
        mLastRotation3D = rotation3D
    }

    override fun draw(
        @Size(2) canvasSizes: IntArray,
        canvas: Canvas,
        scrollRate: Float,
        rotation2D: Float,
        rotation3D: Float,
    ) {
        if (scrollRate < 1) {
            canvas.rotate(
                rotation2D,
                canvasSizes[0] * 0.5f,
                canvasSizes[1] * 0.5f
            )
            for (s in mSnows) {
                mPaint.color = s.color
                mPaint.alpha = ((1 - scrollRate) * 255).toInt()
                canvas.drawCircle(s.centerX, s.centerY, s.radius, mPaint)
            }
        }
    }

    companion object {
        private const val INITIAL_ROTATION_3D = 1000f
        private const val SNOW_COUNT = 50
        private const val SNOW_SPEED = 1.5

        @ColorInt
        fun getThemeColor(daylight: Boolean): Int {
            return if (daylight) -0x974501 else -0xe5a46e
        }
    }
}
